compat/mingw.c

```diff
@@ -475,6 +475,50 @@ static wchar_t *normalize_ntpath(wchar_t *wbuf)
 	return wbuf;
 }
 
+/* rename and unlink a file. This is needed for the ability to recreate a file
+ * with the same name, for example on checkout which unlinks and recreates the file */
+static int rename_and_unlink(const wchar_t *pathname, int (*unlinker)(const wchar_t *))
+{
+	int res;
+	/* leave some space for the .unlink_XXXXXX extension */
+	wchar_t temppath[MAX_LONG_PATH + 15];
+	*temppath = 0;
+
+	wcsncat(temppath, pathname, sizeof(temppath));
+	wcsncat(temppath, L".unlink_XXXXXX", sizeof(temppath));
+	if (!_wmktemp(temppath) || !MoveFileExW(pathname, temppath, 0))
+		return 1;
+	res = unlinker(temppath);
+	if (res) {
+		const DWORD err = GetLastError();
+		MoveFileExW(temppath, pathname, 0);
+		SetLastError(err);
+	}
+	return res;
+}
+
+static int needs_rename_before_unlink(const wchar_t *pathname, HANDLE *handle)
+{
+	/* If the file is opened by another process, rename it before unlinking */
+	*handle = CreateFileW(pathname, GENERIC_WRITE, FILE_SHARE_DELETE, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
```

Comment:
One of the downsides of this approach is that if the file is read-only, this will be indistinguishable from a file that's locked by another process. Subsequently, MoveFileExW will succeed on the file (the +r attribute doesn't affect this), but subsequent delete will fail.

Reply:
Are you sure that a read-only file cannot be deleted? At least on linux, you need to have write permissions on the file's *directory* for deleting, regardless of the file itself.

Reply:
@orgads this ain't linux 🤷🏻‍♂️

I'd suggest to read this comment. It has a great rundown of the issues at play: https://github.com/Aaronepower/remove_dir_all/blob/8f5a95f0736b40adc67263def42fd1aca385393c/src/fs.rs#L34

Reply:
Thank you. That's enlightening. :)

Reply:
@piscisaureus: Ok, I just noticed that the file is given 0666 permissions before calling this function.